packages/edit-site/src/components/block-editor/use-site-editor-settings.js

```diff
@@ -29,18 +34,42 @@ function useNavigateToPreviousEntityRecord() {
 		const didComeFromEditorCanvas =
 			previousLocation?.query.canvas === 'edit';
 		const showBackButton = isFocusMode && didComeFromEditorCanvas;
-		return showBackButton ? () => history.back() : undefined;
+
+		if ( ! showBackButton ) {
+			return undefined;
+		}
+
+		return () => {
+			// Capture currently selected block before navigating back
+			const currentSelectedBlockClientId = getSelectedBlockClientId();
+
+			// Store selected block for current location
+			const currentPath =
+				window.location.pathname + window.location.search;
+			if ( currentSelectedBlockClientId ) {
+				window.sessionStorage?.setItem(
+					`gutenberg_selected_block_${ currentPath }`,
+					currentSelectedBlockClientId
+				);
+			}
+
+			history.back();
+		};
 		// `previousLocation` changes when the component updates for any reason, not
 		// just when location changes. Until this is fixed we can't add it to deps. See
 		// https://github.com/WordPress/gutenberg/pull/58710#discussion_r1479219465.
-	}, [ location, history ] );
+	}, [ location, history, getSelectedBlockClientId ] );
 	return goBack;
 }
 
 export function useSpecificEditorSettings() {
 	const { query } = useLocation();
 	const { canvas = 'view' } = query;
 	const onNavigateToEntityRecord = useNavigateToEntityRecord();
+
+	// Restore block selection when navigating
+	useRestoreBlockSelection();
```

Comment:
Why is this fix edit-site specific. I'd have expected most of the change to happen actually in "editor" package and potentially some small "implementation" changes in the onNavigateToEntity and onNavigateBack callbacks (edit-site and edit-post)